File: internal/context/context.go
package context

import (
	"fmt"
	"os/exec"
	"runtime"
	"strings"
)

// OSContext holds collected system information.
type OSContext struct {
	OS           string
	Arch         string
	Distro       string
	Kernel       string
	Hostname     string
	CurrentUser  string
	ShellVersion string
}

// Collect gathers OS information from the current system.
func Collect() *OSContext {
	ctx := &OSContext{
		OS:   runtime.GOOS,
		Arch: runtime.GOARCH,
	}

	ctx.Distro = runCmd("lsb_release", "-ds")
	if ctx.Distro == "" {
		ctx.Distro = runCmd("cat", "/etc/os-release")
	}
	ctx.Kernel = runCmd("uname", "-r")
	ctx.Hostname = runCmd("hostname")
	ctx.CurrentUser = runCmd("whoami")
	ctx.ShellVersion = runCmd("bash", "--version")

	return ctx
}

// String formats the context for inclusion in LLM prompts.
func (c *OSContext) String() string {
	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("OS: %s/%s\n", c.OS, c.Arch))
	if c.Distro != "" {
		sb.WriteString(fmt.Sprintf("Distro: %s\n", firstLine(c.Distro)))
	}
	if c.Kernel != "" {
		sb.WriteString(fmt.Sprintf("Kernel: %s\n", c.Kernel))
	}
	if c.Hostname != "" {
		sb.WriteString(fmt.Sprintf("Hostname: %s\n", c.Hostname))
	}
	if c.CurrentUser != "" {
		sb.WriteString(fmt.Sprintf("User: %s\n", c.CurrentUser))
	}
	return sb.String()
}

func runCmd(name string, args ...string) string {
	out, err := exec.Command(name, args...).Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

func firstLine(s string) string {
	if idx := strings.IndexByte(s, '\n'); idx >= 0 {
		return s[:idx]
	}
	return s
}

File: nr_diagnose/context.py
"""OS context collection - gathers system information for LLM prompts."""

import platform
import subprocess
from dataclasses import dataclass


@dataclass
class OSContext:
    os_name: str = ""
    arch: str = ""
    distro: str = ""
    kernel: str = ""
    hostname: str = ""
    current_user: str = ""
    shell_version: str = ""

    def to_string(self) -> str:
        """Format the context for inclusion in LLM prompts."""
        parts = [f"OS: {self.os_name}/{self.arch}"]
        if self.distro:
            parts.append(f"Distro: {self.distro.split(chr(10))[0]}")
        if self.kernel:
            parts.append(f"Kernel: {self.kernel}")
        if self.hostname:
            parts.append(f"Hostname: {self.hostname}")
        if self.current_user:
            parts.append(f"User: {self.current_user}")
        return "\n".join(parts)


def _run_cmd(*args: str) -> str:
    """Run a command and return its stripped output, or empty string on failure."""
    try:
        result = subprocess.run(
            args, capture_output=True, text=True, timeout=5
        )
        return result.stdout.strip()
    except (subprocess.SubprocessError, FileNotFoundError, OSError):
        return ""


def collect() -> OSContext:
    """Gather OS information from the current system."""
    ctx = OSContext(
        os_name=platform.system().lower(),
        arch=platform.machine(),
    )

    ctx.distro = _run_cmd("lsb_release", "-ds")
    if not ctx.distro:
        ctx.distro = _run_cmd("cat", "/etc/os-release")

    ctx.kernel = _run_cmd("uname", "-r")
    ctx.hostname = _run_cmd("hostname")
    ctx.current_user = _run_cmd("whoami")
    ctx.shell_version = _run_cmd("bash", "--version")

    return ctx
